<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNA-seq data processing pipeline • rseqR</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="RNA-seq data processing pipeline">
<meta property="og:description" content="An R-based pipeline to analyse RNA-seq data. This pipeline performs
  quality control of reads, generates gene and transcript level counts, performs 
  differential analysis, and gene pathway enrichment analyses.">
<meta property="og:image" content="https://github.com/anilchalisey/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">rseqR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/anilchalisey/rseqR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    

    
    
<div id="rseqr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#rseqr" class="anchor"></a>rseqR <img src="reference/figures/logo.png" align="right">
</h1></div>
<p>The rseqR package is designed to perform differential analysis of RNA-seq data. It provides tools to:</p>
<ul>
<li>perform quality control of raw reads in fastq format and generate HTML reports (using FASTQC and MULTIQC)</li>
<li>determine trancript and gene-level read count estimates (using Salmon)</li>
<li>run differential analysis using DESeq2, edgeR, and Limma-voom and to export the results into easily readable tab-delimited files</li>
<li>perform pathway analysis of differentially expressed genes</li>
</ul>
<p>The rseqR package does not attempt to replace the tools it uses, but instead provides an automated environment in which to use them.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div id="command-line-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#command-line-tools" class="anchor"></a>Command line tools</h2>
<p>The workflow used by rseqR requires the installation of several command-line tools: <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Fastqc%5D">Fastqc v0.11.7</a>, <a href="http://multiqc.info/">MultiQC v1.6</a>, and <a href="https://combine-lab.github.io/salmon/">Salmon</a>. Installation instructions for these may be found at their respective websites, but a guide is given below for convenience.</p>
<p>If using a Unix-based system, open up a terminal and follow the commands as is. If using the windows subsystem for linux (WSL) on Windows 10 then WSL must first be set up as detailed <a href="https://docs.microsoft.com/en-us/windows/wsl/install-win10">here</a>. Once WSL is up and running, then the tools may be installed as on any Unix-based system.</p>
<p>It is assumed that root priviliges are available (if necessary) - if not, then a system administrator may need to install these for you.</p>
<div id="fastqc" class="section level3">
<h3 class="hasAnchor">
<a href="#fastqc" class="anchor"></a>Fastqc</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="fu">wget</span> https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.7.zip</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="fu">unzip</span> fastqc_v0.11.7.zip</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="bu">cd</span> FastQC</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="fu">chmod</span> 755 fastqc</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="bu">cd</span> /usr/local/bin</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="fu">sudo</span> ln -s ~/Fastqc/fastqc .</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="fu">ln</span> -s ~/Fastqc/fastqc .</a></code></pre></div>
</div>
<div id="multiqc" class="section level3">
<h3 class="hasAnchor">
<a href="#multiqc" class="anchor"></a>MultiQC</h3>
<p>To run MultiQC ensure there is a working python distribution.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="ex">pip</span> install multiqc</a></code></pre></div>
</div>
<div id="salmon" class="section level3">
<h3 class="hasAnchor">
<a href="#salmon" class="anchor"></a>Salmon</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="fu">wget</span> https://github.com/COMBINE-lab/salmon/releases/download/v0.11.2/salmon-0.11.2-linux_x86_64.tar.gz</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="fu">tar</span> -xvzf salmon-0.11.2-linux_x86_64.tar.gz</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co"># If using WSL</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="bu">cd</span> /usr/local/bin </a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="fu">sudo</span> ln -s ~/salmon-0.11.2-linux_x86_64/bin/salmon .</a>
<a class="sourceLine" id="cb3-8" data-line-number="8"></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co"># If using other Unix</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="bu">cd</span> ~/bin</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="fu">ln</span> -s ~/salmon-0.11.2-linux_x86_64/bin/salmon .</a></code></pre></div>
</div>
<div id="r-package-dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#r-package-dependencies" class="anchor"></a>R package dependencies</h3>
<p>There are a number of dependencies to the rseqR package detailed in the <code>DESCRIPTION</code> file. These may be installed manually, otherwise the following commands may be used:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">pkgs &lt;-<span class="st"> </span><span class="kw">installed.packages</span>()[, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb4-2" data-line-number="2">install.from.cran &lt;-<span class="st"> </span>pkgs[<span class="op">!</span>(<span class="kw">c</span>(<span class="st">"utils"</span>, <span class="st">"parallel"</span>, <span class="st">"dplyr"</span>, <span class="st">"data.table"</span>, <span class="st">"purrr"</span>, <span class="st">"stats"</span>,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3">                              <span class="st">"grDevices"</span>, <span class="st">"ggplot2"</span>, <span class="st">"ggrepel"</span>, <span class="st">"doParallel"</span>, <span class="st">"magrittr"</span>,</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">                              <span class="st">"ggdendro"</span>, <span class="st">"viridis"</span>, <span class="st">"grid"</span>, <span class="st">"gridExtra"</span>, <span class="st">"gtable"</span>, <span class="st">"tools"</span>,</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">                              <span class="st">"crayon"</span>, <span class="st">"UpSetR"</span>, <span class="st">"devtools"</span>, <span class="st">"data.tree"</span>) <span class="op">%in%</span><span class="st"> </span>pkgs)]</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">install.from.bc &lt;-<span class="st"> </span>pkgs[<span class="op">!</span>(<span class="kw">c</span>(<span class="st">"tximport"</span>, <span class="st">"edgeR"</span>, <span class="st">"DESeq2"</span>, <span class="st">"limma"</span>, <span class="st">"gage"</span>, <span class="st">"org.Hs.eg.db"</span>, </a>
<a class="sourceLine" id="cb4-7" data-line-number="7">                            <span class="st">"org.Mm.eg.db"</span>, <span class="st">"SummarizedExperiment"</span>) <span class="op">%in%</span><span class="st"> </span>pkgs)]</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="kw">install.packages</span>(install.from.cran)</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="kw">biocLite</span>(install.from.bc)</a></code></pre></div>
</div>
<div id="install-rseqr" class="section level3">
<h3 class="hasAnchor">
<a href="#install-rseqr" class="anchor"></a>Install rseqR</h3>
<p>Once all dependencies are installed, then rseqR may be installed as follows:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"anilchalisey/rseqr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>The entire workflow of rseqR may be run by a call to a single function <code>run_dea()</code>, as shown below.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw">run_dea</span>(<span class="dt">sample.info =</span> <span class="st">`</span><span class="dt">system.file("extdata", "paired-example.txt", pkg = "rseqr")</span><span class="st">`</span>,</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">                  <span class="dt">reference =</span> <span class="kw">c</span>(<span class="st">"WT"</span>, <span class="st">"KO1"</span>, <span class="st">"KO2"</span>), <span class="dt">species =</span> <span class="st">"human"</span>, <span class="dt">output.dir =</span> <span class="st">"results-dea"</span>,</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">                  <span class="dt">threads =</span> <span class="ot">NULL</span>, <span class="dt">index.dir =</span> <span class="ot">NULL</span>)</a></code></pre></div>
<p>The arguments to the <code>run_dea()</code> function are:</p>
<div id="sample-info" class="section level3">
<h3 class="hasAnchor">
<a href="#sample-info" class="anchor"></a>sample.info:</h3>
<p>This is the path to a tab-delimited file with at least the columns:</p>
<ul>
<li>
<code>condition</code>: treatment or condition labels on the basis of which the differential analysis will be performed (e.g. “CONTROL”, “TREATED”; or “WILD-TYPE”, “KNOCK-OUT1”, “KNOCK-OUT2”).</li>
<li>
<code>sample</code>: names of the samples - these labels will be used in plots and as sample column headers in the output.</li>
<li>
<code>file1</code>: absolute or relative path to fastq or salmon quant.sf files.</li>
</ul>
<p>Optional columns include:</p>
<ul>
<li>
<code>file2</code>: if fastq files and PE reads, then this column should also be present, specifying the absolute or relative paths to the second pair of fastq reads.<br>
# <code>batch</code>: if a batch effect is to be included in the design, then this should be identified under this column (e.g. litter number or sequencing run).</li>
</ul>
<p>If starting from fastq files, then the workflow will include QC of raw reads, and read count estimation with salmon. If a salmon index has not been identified then an index will be built automatically prior to read count estimation. If starting from quant.sf files, then the analysis will start directly at the differential analysis stage.</p>
<p>Dummy examples of the tab-delimited files accepted by rseqR may be found at <code>system.file("extdata", "paired-example.txt", pkg = "rseqr")</code>, <code>system.file("extdata", "sinle-example.txt", pkg = "rseqr")</code>, and <code>system.file("extdata", "quants-example.txt", pkg = "rseqr")</code>.</p>
</div>
<div id="reference" class="section level3">
<h3 class="hasAnchor">
<a href="#reference" class="anchor"></a>reference:</h3>
<p>The order in which the condition labels should be evaluated when performing differential analysis. For example, c(“A”, “B”, “C”, “D”) would mean “A” is the reference condition to which “B”, “C” and “D” are compared; in addition, “C” and “D” will be compared to “B”, and “D” will be compared to “C”. If  then the comparisons will be arranged alphabetically.</p>
</div>
<div id="species" class="section level3">
<h3 class="hasAnchor">
<a href="#species" class="anchor"></a>species:</h3>
<p>May be one of “human” or “mouse” - other options will be added in later versions.</p>
</div>
<div id="output-dir" class="section level3">
<h3 class="hasAnchor">
<a href="#output-dir" class="anchor"></a>output.dir:</h3>
<p>Directory to which results should be saved.</p>
</div>
<div id="threads" class="section level3">
<h3 class="hasAnchor">
<a href="#threads" class="anchor"></a>threads:</h3>
<p>Number of threads to be used when parallelisation is possible. If <code>NULL</code> then one less than the maximum numbre of threads available will be used.</p>
</div>
<div id="index-dir" class="section level3">
<h3 class="hasAnchor">
<a href="#index-dir" class="anchor"></a>index.dir:</h3>
<p>The path to the salmon index.</p>
<p>Other arguments are also possible, and details for these may be found in the manual, but for most users, the default settings are satisfactory.</p>
</div>
</div>
<div id="output" class="section level2">
<h2 class="hasAnchor">
<a href="#output" class="anchor"></a>Output</h2>
<p>Once complete, all the results will be saved in the specified output directory with the following structure:</p>
<pre><code>#&gt;                levelName
#&gt; 1  results              
#&gt; 2   ¦--DE_results.rda   
#&gt; 3   ¦--DESeq2           
#&gt; 4   ¦--edgeR            
#&gt; 5   ¦--fastqc           
#&gt; 6   ¦--index            
#&gt; 7   ¦--Limma-voom       
#&gt; 8   ¦--metadata.rda     
#&gt; 9   ¦--multiqc          
#&gt; 10  ¦--salmon           
#&gt; 11  ¦--upset_KO1-WT.png 
#&gt; 12  ¦--upset_KO2-WT.png 
#&gt; 13  °--upset_KO2-KO1.png</code></pre>
<p>The directories <code>DESeq2</code>, <code>edgeR</code>, <code>fastqc</code>, <code>Limma-voom</code>, and <code>multiqc</code> contain the results of the respective analyses including relevant plots. Within the <code>salmon</code> directory one will find individual directories for the quantification results of each sample. The <code>index</code> directory contains the salmon index for human or mouse if it had not been created and specified beforehand. <code>DE-results.rda</code> contains the results of the differential expression analysis and comprises a list of three objects: <code>DE_limma</code>, <code>DE_edger</code>, and <code>DE_deseq2</code>. The <code>metadata.rda</code> object contains the metadata for the samples including the count data for each sample; this may be used to re-run the analysis from an intermediate step.</p>
</div>
<div id="running-only-part-of-the-rseqr-pipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#running-only-part-of-the-rseqr-pipeline" class="anchor"></a>Running only part of the rseqR pipeline</h2>
<p>If there is a need to reanalyse only a portion of the data, then this may be done using the <code>metadata.rda</code> object, which contains all the necessary information. For example, it is possible to run each of the differential analysis tools separately using the commands <code>limma_voom_analysis()</code>, <code>edger_analysis()</code>, or <code>deseq2_analysis()</code>. For example:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">load</span>(<span class="st">"results/metadata.rda"</span>) <span class="co"># this will load an object called 'metadata' into the environment</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">DE_limma &lt;-<span class="st"> </span><span class="kw">limma_voom_analysis</span>(<span class="dt">metadata =</span> metadata, <span class="dt">species =</span> <span class="st">"human"</span>)</a></code></pre></div>
</div>
<div id="pathway-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#pathway-analysis" class="anchor"></a>Pathway analysis</h2>
<p>Gene ontology and KEGG pathway analysis may be performed using the <code>gage</code> package on the objects contained within <code>DE-results.rda</code>. Loading this RDA file will populate the environment with the objects <code>DE_limma</code>, <code>DE_edger</code>, and <code>DE_deseq2</code>. Each of these is a list with the results of pairwise differential expression analyses. In the example used abvoe, the three conditions were <code>WT</code>, <code>KO1</code>, and <code>KO2</code>; thus, the three comparisons are <code>KO1-WT</code>, <code>KO2-WT</code>, and <code>KO2-KO1</code>. To perform pathway analysis on the differential genes identified by DESeq2 for the first comparison, the following command may be used:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">pa1_deseq2 &lt;-<span class="st"> </span><span class="kw">pathway_analysis</span>(DE_deseq2[[<span class="dv">1</span>]], <span class="dt">species =</span> <span class="st">"human"</span>)</a></code></pre></div>
<p>This may be repeated as necessary for the other comparisons.</p>
</div>
<div id="about-rseqr" class="section level2">
<h2 class="hasAnchor">
<a href="#about-rseqr" class="anchor"></a>About rseqR</h2>
<p>The rseqR package has been developed in the Chris O’Callaghan Group at the Centre for Cellular and Molecular Biology, University of Oxford by Anil Chalisey and Chris O’Callaghan.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/anilchalisey/rseqR">https://​github.com/​anilchalisey/​rseqR</a>
</li>
<li>Report a bug at <br><a href="https://github.com/anilchalisey/rseqR/issues">https://​github.com/​anilchalisey/​rseqR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/anilchalisey">Anil Chalisey</a> <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/anilchalisey">Anil Chalisey</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
