<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Quantify transcript abundances using Salmon — run_salmon • rseqR</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Quantify transcript abundances using Salmon — run_salmon" />

<meta property="og:description" content="Run the abundance quantification tool Salmon on a set of FASTQ
files. Requires Salmon (https://combine-lab.github.io/salmon/)
to be installed and a Salmon transcript index must have been generated prior
to using this function. See the Salmon website for installation and basic
usage instructions." />

<meta property="og:image" content="https://anilchalisey.github.io/rseqR/logo.png" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rseqR</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/anilchalisey/rseqR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Quantify transcript abundances using Salmon</h1>
    <small class="dont-index">Source: <a href='https://github.com/anilchalisey/rseqR/blob/master/R/run_salmon.R'><code>R/run_salmon.R</code></a></small>
    <div class="hidden name"><code>run_salmon.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Run the abundance quantification tool <code>Salmon</code> on a set of FASTQ
files. Requires <code>Salmon</code> (<a href='https://combine-lab.github.io/salmon/'>https://combine-lab.github.io/salmon/</a>)
to be installed and a Salmon transcript index must have been generated prior
to using this function. See the Salmon website for installation and basic
usage instructions.</p>
    
    </div>

    <pre class="usage"><span class='fu'>run_salmon</span>(<span class='no'>fastq1</span>, <span class='kw'>fastq2</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>index.dir</span>, <span class='kw'>dest.dir</span> <span class='kw'>=</span> <span class='st'>"SALMON"</span>,
  <span class='kw'>salmon</span> <span class='kw'>=</span> <span class='st'>"salmon"</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>advanced.opts</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bam</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>bootstraps</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>seqBias</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>gcBias</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>posBias</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>allowOrphans</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fastq1</th>
      <td><p>a character vector indicating the read files to be trimmed.</p></td>
    </tr>
    <tr>
      <th>fastq2</th>
      <td><p>(optional) a character vector indicating read files to be
trimmmed.  If specified, it is assumed the reads are paired, and this vector
MUST be in the same order as those listed in <code>fastq1</code>.  If <code>NULL</code>
then it is assumed the reads are single-end.</p></td>
    </tr>
    <tr>
      <th>index.dir</th>
      <td><p>directory of the index files needed for read mapping using Salmon.
See function <code>'build_index()'</code>.</p></td>
    </tr>
    <tr>
      <th>dest.dir</th>
      <td><p>directory where results are to be saved.  If directory does not exist,
then it will be created.</p></td>
    </tr>
    <tr>
      <th>salmon</th>
      <td><p>(optional) string giving full command to use to call
Salmon, if simply typing "salmon" at the command line does not give the
required version of Salmon or does not work. Default is simply "salmon".
If used, this argument should give the full path to the desired Salmon
binary.</p></td>
    </tr>
    <tr>
      <th>threads</th>
      <td><p>an integer value indicating the number of parallel threads to
be used by FastQC. [DEFAULT = maximum number of available threads - 1].</p></td>
    </tr>
    <tr>
      <th>advanced.opts</th>
      <td><p>character vector supplying list of advanced option
arguments to apply to each Salmon call. For details see Salmon documentation
or type <code>salmon quant --help-reads</code> at the command line.</p></td>
    </tr>
    <tr>
      <th>bam</th>
      <td><p>logical, if <code>TRUE</code> then create a pseudo-alignment BAM file.
[Default = <code>FALSE</code>]</p></td>
    </tr>
    <tr>
      <th>bootstraps</th>
      <td><p>integer giving the number of bootstrap samples
that Salmon should use (default is 0). With bootstrap samples, uncertainty
in abundance can be quantified.</p></td>
    </tr>
    <tr>
      <th>seqBias</th>
      <td><p>logical, should Salmon's option be used to model and correct
abundances for sequence specific bias? Default is <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>gcBias</th>
      <td><p>logical, should Salmon's option be used to model and correct
abundances for GC content bias? Requires Salmon version 0.7.2 or higher.
Default is <code>TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>posBias</th>
      <td><p>logical, should Salmon's option be used to model and correct
abundances for positional biases? Requires Salmon version 0.7.3 or higher.
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>allowOrphans</th>
      <td><p>logical, if <code>TRUE</code> then consider orphaned reads as
valid hits when performing lightweight-alignment. This option will increase
sensitivity (allow more reads to map and more transcripts to be detected), but may
decrease specificity as orphaned alignments are more likely to be spurious.
For more details see Salmon documentation.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The following items will be returned and saved in the salmon directory:</p><ol>
<li><p>quant.sf: plain-text, tab-separated quantification file
that contains 5 column: Name,Length,EffectiveLength,TPM, and NumReads.</p></li>
<li><p>quant.sf.bkp: plain-text, tab-separated quantification file
that contains 5 column: Name,Length,EffectiveLength,TPM, and NumReads.
This is the raw version of the quant.sf file.</p></li>
<li><p>cmd_info.json: A JSON format file that records the main command
line parameters with which Salmon was invoked for the run that produced
the output in this directory.</p></li>
<li><p>aux_info: This directory will have a number of files (and
subfolders) depending on how salmon was invoked.</p></li>
<li><p>meta_info.json: A JSON file that contains meta information about
the run, including stats such as the number of observed and mapped
fragments, details of the bias modeling etc.</p></li>
<li><p>ambig_info.tsv: This file contains information about the number
of uniquely-mapping reads as well as the total number of
ambiguously-mapping reads for each transcript.</p></li>
<li><p>lib_format_counts.json: This JSON file reports the number of
fragments that had at least one mapping compatible with the designated
library format, as well as the number that didn't.</p></li>
<li><p>libParams: The auxiliary directory will contain a text file
called flenDist.txt. This file contains an approximation of the
observed fragment length distribution.</p></li>
</ol>

    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Rob Patro, Geet Duggal, Michael I. Love, Rafael A. Irizarry, and
Carl Kingsford (2017): Salmon provides fast and bias-aware
quantification of transcript expression. Nature methods, 14(4), 417.
<a href='https://www.nature.com/articles/nmeth.4197'>https://www.nature.com/articles/nmeth.4197</a></p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#references">References</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://github.com/anilchalisey'>Anil Chalisey</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

